src/number_theory/divisors.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Aaron Anderson
+-/
+import data.nat.totient
+import number_theory.quadratic_reciprocity
+
+/-!
+# Divisor finsets
+
+This file defines sets of divisors of a natural number. This is particularly useful as background
+for defining Dirichlet convolution.
+
+## Main Definitions
+Let `n : ℕ`. All of the following definitions are in the `nat` namespace:
+ * `divisors n` is the `finset` of natural numbers that divide `n`.
+ * `proper_divisors n` is the `finset` of natural numbers that divide `n`, other than `n`.
+ * `divisors_antidiagonal n` is the `finset` of pairs `(x,y)` such that `x * y = n`.
+ * `perfect n` is true when the sum of `proper_divisors n` is `n`.
+
+## Implementation details
+ * All of `divisors 0`, `proper_divisors 0`, and `divisors_antidiagonal 0` are defined to be `∅`.
+
+## Tags
+divisors, perfect numbers
+
+-/
+
+open_locale classical
+open_locale big_operators
+
+namespace nat
+variable (n : ℕ)
+
+/-- `divisors n` is the `finset` of divisors of `n`. As a special case, `divisors 0 = ∅`. -/
+def divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.Ico 1 (n + 1))
+
+/-- `proper_divisors n` is the `finset` of divisors of `n`, other than `n`.
+  As a special case, `divisors 0 = ∅`. -/
+def proper_divisors : finset ℕ := finset.filter (λ x : ℕ, x ∣ n) (finset.Ico 1 n)
+
+/-- `divisors_antidiagonal n` is the `finset` of pairs `(x,y)` such that `x * y = n`.
+  As a special case, `divisors_antidiagonal 0 = ∅`. -/
+def divisors_antidiagonal : finset (ℕ × ℕ) :=
+((finset.Ico 1 (n + 1)).product (finset.Ico 1 (n + 1))).filter (λ x, x.fst * x.snd = n)
+
+variable {n}
+
+lemma not_proper_self : ¬ n ∈ proper_divisors n :=
```

Comment:
`nat.not_proper_self` doesn't really spell out the meaning of this lemma. How about:

```suggestion
lemma proper_divisors.not_self_mem : ¬ n ∈ proper_divisors n :=
```